kernel/kernel-defaults/src/test/scala/io/delta/kernel/defaults/DeltaTableSchemaEvolutionSuite.scala

```diff
@@ -0,0 +1,1249 @@
+    withTempDirAndEngine { (tablePath, engine) =>
+      val table = Table.forPath(engine, tablePath)
+      val initialSchema = new StructType()
+        .add("a", StringType.STRING, true)
+      createEmptyTable(
+        engine,
+        tablePath,
+        initialSchema,
+        tableProperties = Map(
+          TableConfig.COLUMN_MAPPING_MODE.getKey -> "id",
+          TableConfig.ICEBERG_COMPAT_V2_ENABLED.getKey -> "true"))
+
+      val currentSchema = table.getLatestSnapshot(engine).getSchema()
+
+      val newSchema = new StructType()
+        .add("a", StringType.STRING, true, currentSchema.get("a").getMetadata)
+        .add(
+          "array_of_arrays",
+          new ArrayType(new ArrayType(IntegerType.INTEGER, true), true),
+          true,
+          FieldMetadata.builder()
+            .putString(ColumnMapping.COLUMN_MAPPING_PHYSICAL_NAME_KEY, "array_of_arrays")
+            .putLong(ColumnMapping.COLUMN_MAPPING_ID_KEY, 4L)
+            .putFieldMetadata(
+              ColumnMapping.COLUMN_MAPPING_NESTED_IDS_KEY,
+              FieldMetadata.builder().putLong("array_of_arrays.element", 2L)
+                .putLong("array_of_arrays.element.element", 3L).build()).build())
+
+      table.createTransactionBuilder(engine, testEngineInfo, Operation.MANUAL_UPDATE)
+        .withSchema(engine, newSchema)
+        .build(engine).commit(engine, emptyIterable())
+
+      val updatedSchema = table.getLatestSnapshot(engine).getSchema()
+
+      assertColumnMapping(updatedSchema.get("a"), 1)
+      assertColumnMapping(updatedSchema.get("array_of_arrays"), 4L, "array_of_arrays")
+
+      val arrayMetadata = updatedSchema.get("array_of_arrays").getMetadata
+      assert(arrayMetadata.getMetadata(ColumnMapping.COLUMN_MAPPING_NESTED_IDS_KEY)
+        .getLong("array_of_arrays.element") == 2L)
+      assert(arrayMetadata.getMetadata(ColumnMapping.COLUMN_MAPPING_NESTED_IDS_KEY)
+        .getLong("array_of_arrays.element.element") == 3L)
+    }
+  }
+
+  test("Changing column mapping on table and evolve schema at same time fails") {
+    withTempDirAndEngine { (tablePath, engine) =>
+      val table = Table.forPath(engine, tablePath)
+      val initialSchema = new StructType()
+        .add("a", StringType.STRING, true)
+        .add("c", IntegerType.INTEGER, true)
+
+      // Create a table initially without column mapping
+      createEmptyTable(engine, tablePath, initialSchema)
+
+      val currentSchema = table.getLatestSnapshot(engine).getSchema()
+      val newSchema = new StructType()
+        .add("a", StringType.STRING, true, currentSchema.get("a").getMetadata)
+        .add(
+          "b",
+          new StructType()
+            .add("d", IntegerType.INTEGER, true, fieldMetadataForColumn(4, "d"))
+            .add("e", IntegerType.INTEGER, true, fieldMetadataForColumn(5, "e")),
+          true,
+          fieldMetadataForColumn(3, "b"))
+        .add("c", IntegerType.INTEGER, true, currentSchema.get("c").getMetadata)
```

Comment:
why don't we just use currentSchema.add(...)...?

Reply:
We can but I liked it better for these tests to explicitly be able to see the before and after.